File: 94. Binary Tree Inorder Traversal.py
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def inorderTraversal_iterative(self, root: TreeNode) -> List[int]:
        ##iterative method :
        l=[]
        stack=[]
        current = root
        while True:
            if current != None:
                stack.append(current)  ##Keep pushing Nodes inside the stack.
                current = current.left
            elif stack:
                current = stack.pop()  ##Take out the nodes one by one from stack.
                l.append(current.val)
                current = current.right
            else:
                break
        return l

    def inorderTraversal_recursive(self, root: TreeNode) -> List[int]:
        ##recursive method :
        if root is None:
            return []
        l = self.inorderTraversal_recursive(root.left)
        r = self.inorderTraversal_recursive(root.right)
        return l + [root.val] + r

File: 94. Binary Tree Inorder Traversal.c
void travel(struct TreeNode* root,int*ret,int*size){
    if(root==NULL)
        return 0 ;
    travel(root->left,ret,size);
    ret[(*size)++]=root->val;
    travel(root->right,ret,size);
}

int* inorderTraversal(struct TreeNode* root, int* returnSize){
    *returnSize=0;
    int *ret=malloc(sizeof(int)*100);
    travel(root,ret,returnSize);
        return ret;

}
